src/input/mod.rs

```diff
@@ -327,15 +327,51 @@ impl State {
                 )
             },
         ) else {
+            self.remove_all_bind_repeat_timers();
             return;
         };
 
-        // Filter actions when the key is released or the session is locked.
-        if !pressed {
-            return;
+        if pressed {
+            self.handle_bind(bind.clone());
+
+            match self.niri.bind_repeat_timers.entry(bind.key) {
+                Entry::Vacant(entry) => {
+                    let repeat_delay_timer = Timer::from_duration(Duration::from_millis(
+                        self.niri.config.borrow().input.keyboard.repeat_delay as u64,
+                    ));
+
+                    let token = self
+                        .niri
+                        .event_loop
+                        .insert_source(repeat_delay_timer, move |_, _, state| {
+                            state.handle_bind(bind.clone());
+
+                            TimeoutAction::ToDuration(Duration::from_millis(max(
+                                1000 / state.niri.config.borrow().input.keyboard.repeat_rate as u64,
+                                100,
+                            )))
+                        })
+                        .unwrap();
+
+                    entry.insert(token);
+                }
+                Entry::Occupied(_) => {
+                    // already repeating, nothing to do
+                }
+            }
+        } else {
+            self.remove_all_bind_repeat_timers();
         }
+    }
 
-        self.handle_bind(bind);
+    fn remove_all_bind_repeat_timers(&mut self) {
```

Comment:
At this point instead of a HashMap, might as well just have a single Option<(Key, Timer)> field.